### ISSUE TITLE
Add `prefix:` option to links created with `glob: true`

### ISSUE THREAD

**Issue by eengstrom** [anishathalye/dotbot#259]
Perhaps this is a niche case, but in my `dotfiles`, I have a **subdirectory** for most of the true "dot" files.  i.e.:

```
dotfiles
├── README.md
├── bin
│   ├── dotbot
│   ├── look
│   ├── pbfile
│   └── ...
├── dot
│   ├── bashrc
│   ├── gitconfig
│   ├── gitignore
│   ├── gorc
│   ├── login
│   ├── ...
│   ├── zshrc
│   └── zshenv
├── ...
```

I don't believe there is a way using the `link: ` module to create symlinks in my home directory for **all** the files in the `dot` subdirectory, sans excluded ones, given #248 when merged.

That is, what I want is to replace the MANY lines in my `dotconf.yml` like this:

```
- link:
    # Then everything in dot subdir:
    # Wish I could do this via a glob
    ~/.bashrc:          dot/bashrc
    ~/.gitconfig:       dot/gitconfig
    ~/.gitignore:       dot/gitignore
   ...
```
... with a glob version, something like this:

```
- link:
    ~/:
       glob: true
       path: ~/.dotfiles/dot/*
       prefix: '.'
```

Note the addition of the `prefix: ` option which specifies the prefix to be added to the basename of the file to be symlinked.  

Alternatives welcome, or if I missed something in the documentation for how this should be configured, a pointer to the documentation.

**Comment by etkeys** [comment]
I too ran into this problem (#244 and PR #245). I'm testing with a patch version of `dotbot` that includes this PR with just to convert over to dotbot. Here is what [my conf file](https://github.com/etkeys/dotfiles/blob/wip/dotbot/master-install.conf.yaml#L108-L110) looks like with this patch. The files in my repo all begin with '.'. Granted it still requires two entries under `link` if I wanted to link files that are prefixed with '.' and also those that are not.

I opted to go this route because it leaves the original uses of `glob` unaffected and made the most sense at the time. I also didn't want to deal with the nuances that the [glob documentation](https://docs.python.org/3.6/library/glob.html) hints at (#lazy). Yes, looking back on it now, there are other routes to try and would still not hurt backwards compatibility. Having a flag parameter to do a glob search for items without a leading '.' and then do second round search for items with a leading '.' would more appealing but might obscure "explicitness".

**Comment by eengstrom** [comment]
thanks @etkeys - I saw that issue/PR, but I think my case is the dual of that.  In my case, none of my config files have leading `.` when stored in my git repo.  Instead, I want to __add__ the leading period when the links are created via `dotbot`.  Ideally, I would **NOT** have to store them in git with leading periods, as that makes editing / etc more difficult, IMO.  

Separately, thanks for sharing your config file example - my goodness, there's quite a lot in there.  In our environment, which I lead and manage, we use `ansible` to configure the system.  I'm looking at `dotbot` just for my personal configuration.    If yours is for personal system config, that might be overkill, but if you start to manage many systems, even just a small handful, I really would look at `ansible`.

**Comment by anishathalye** [comment]
This closely mirrors Dotbot's default behavior when linking explicitly listed files in the base directory:

> If the source location is omitted or set to null, Dotbot will use the basename of the destination, with a leading . stripped if present.

So it does seem reasonable to support this sort of functionality when using glob.

Is there any case where someone would use a prefix other than `.`?

**Comment by eengstrom** [comment]
I can't come up with an **obvious** prefix other than `.`, but figured if the support for a prefix was to be added, having the default be `.`, but ability to specify otherwise, might help someone down the line.  

IOW, all I selfishly care about is adding a leading `.` so I can use `glob` instead of listing each file.

**Comment by eengstrom** [comment]
OTOH, If we were to assume a prefix of `.`, that would disallow linking files that are not true "dot-files".  For example, if I had a collection of directories I wanted symlinked into $HOME, but as is, then the current behavior of `glob` is just right.

I think that means we would need to add a `prefix:` argument, rather than change any form of default behavior.

Thoughts?

**Comment by anishathalye** [comment]
Yes, I agree that the current behavior should be preserved (so if prefix as described above is added, it should default to empty string).

**Comment by eengstrom** [comment]
@anishathalye - so I found a couple of hours to add the functionality.  I've been running with it for a while, but just in the last day I also added test cases, which found some other edge cases.  

I made three separate PRs because I wanted to show that each change is small to get here.  Mostly, I had to fix the globbing assumptions that globs contained `*`, but while I did that, I also allowed it to support other glob pattern symbols (e.g. `?` and `[seq]`)

Let me know if I didn't follow your style or whatever.  Will fix, but I think this is going to be generally useful additions.
